commands/command_filter_process.go

```diff
@@ -40,33 +44,83 @@ func filterCommand(cmd *cobra.Command, args []string) {
 		ExitWithError(err)
 	}
 
-	_, err := s.NegotiateCapabilities()
+	caps, err := s.NegotiateCapabilities()
 	if err != nil {
 		ExitWithError(err)
 	}
 
+	var supportsDelay bool
+	for _, cap := range caps {
+		if cap == "capability=delay" {
+			supportsDelay = true
+			break
+		}
+	}
+
 	skip := filterSmudgeSkip || cfg.Os.Bool("GIT_LFS_SKIP_SMUDGE", false)
 	filter := filepathfilter.New(cfg.FetchIncludePaths(), cfg.FetchExcludePaths())
 
+	ptrs := make(map[string]*lfs.Pointer)
+
+	var q *tq.TransferQueue
+	closeOnce := new(sync.Once)
+	available := make(chan *tq.Transfer)
+
+	if supportsDelay {
+		q = tq.NewTransferQueue(tq.Download, getTransferManifest(), cfg.CurrentRemote)
+		go infiniteTransferBuffer(q, available)
+	}
+
 	var malformed []string
 	var malformedOnWindows []string
-
 	for s.Scan() {
 		var n int64
 		var err error
+		var delayed bool
 		var w *git.PktlineWriter
 
 		req := s.Request()
 
-		s.WriteStatus(statusFromErr(nil))
+		if !(req.Header["command"] == "smudge" && req.Header["can-delay"] == "1") && !(req.Header["command"] == "list_available_blobs") {
+			s.WriteStatus(statusFromErr(nil))
+		}
 
 		switch req.Header["command"] {
 		case "clean":
 			w = git.NewPktlineWriter(os.Stdout, cleanFilterBufferCapacity)
 			err = clean(w, req.Payload, req.Header["pathname"], -1)
 		case "smudge":
 			w = git.NewPktlineWriter(os.Stdout, smudgeFilterBufferCapacity)
-			n, err = smudge(w, req.Payload, req.Header["pathname"], skip, filter)
+			if req.Header["can-delay"] == "1" {
+				var ptr *lfs.Pointer
+
+				n, delayed, ptr, err = delayedSmudge(s, w, req.Payload, q, req.Header["pathname"], skip, filter)
+
+				if delayed {
+					ptrs[req.Header["pathname"]] = ptr
+				}
+			} else {
+				from, err := incomingOrCached(req.Payload, ptrs[req.Header["pathname"]])
+				if err != nil {
+					break
+				}
+
+				n, err = smudge(w, from, req.Header["pathname"], skip, filter)
+
+				if err == nil {
+					delete(ptrs, req.Header["pathname"])
+				}
+			}
+		case "list_available_blobs":
+			closeOnce.Do(func() { go q.Wait() })
```

Comment:
I could use some clarification here. It's my understanding that a transfer queue can accept new transfers until `Wait()` is called. If I'm interpreting [the protocol docs](https://github.com/git/git/commit/51b8aecabed1363f13c990320a50e1fb2aa3f696) correctly, then the process should look like this:

1. Git sends all pointers through `smudge` with `can-delay=1`
2. Git sends `list_available_blobs`. The transfer queue is closed.
3. Git sends more `smudge` commands for each available blob, _never_ calling `list_available_blobs` again.
4. Profit!!